File: problem/ai-battle-series/11-course-rush/src/main.py
import sys


def main():
    data = sys.stdin.read().strip().split()
    if len(data) < 2:
        return
    dept = data[0]
    threshold = int(data[1])

    courses = [
        {
            "name": "進階機器學習",
            "professor": "陳教授",
            "title": "教授",
            "capacity": 30,
            "enrolled": 28,
            "dept": "CS"
        },
        {
            "name": "資料結構",
            "professor": "王教授",
            "title": "副教授",
            "capacity": 50,
            "enrolled": 45,
            "dept": "CS"
        },
        {
            "name": "計算機組織",
            "professor": "李教授",
            "title": "助理教授",
            "capacity": 40,
            "enrolled": 12,
            "dept": "CS"
        },
        {
            "name": "微積分",
            "professor": "張教授",
            "title": "教授",
            "capacity": 60,
            "enrolled": 50,
            "dept": "MATH"
        },
    ]

    results = []
    for c in courses:
        if c["dept"] != dept:
            continue
        popularity = c["enrolled"] * 100.0 / c["capacity"]
        if popularity + 1e-9 < threshold:
            continue
        results.append((popularity, c))

    if not results:
        print("No matching courses")
        return

    results.sort(key=lambda x: (-x[0], x[1]["name"]))
    for popularity, c in results:
        print(
            f"{c['name']} by {c['professor']} ({c['title']}) - {popularity:.1f}% ({c['enrolled']}/{c['capacity']})"
        )


if __name__ == "__main__":
    main()

File: problem/ai-battle-series/11-course-rush/src/main.cpp
#include <bits/stdc++.h>
using namespace std;

struct Course {
    string name;
    string professor;
    string title;
    int capacity;
    int enrolled;
    string dept;
};

struct Result {
    const Course *course;
    double popularity;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string dept;
    int threshold = 0;
    if (!(cin >> dept)) {
        return 0;
    }
    if (!(cin >> threshold)) {
        return 0;
    }

    vector<Course> courses = {
        {"進階機器學習", "陳教授", "教授", 30, 28, "CS"},
        {"資料結構", "王教授", "副教授", 50, 45, "CS"},
        {"計算機組織", "李教授", "助理教授", 40, 12, "CS"},
        {"微積分", "張教授", "教授", 60, 50, "MATH"},
    };

    vector<Result> results;
    for (const auto &course : courses) {
        if (course.dept != dept) {
            continue;
        }
        double popularity = course.enrolled * 100.0 / course.capacity;
        if (popularity + 1e-9 < threshold) {
            continue;
        }
        results.push_back({&course, popularity});
    }

    if (results.empty()) {
        cout << "No matching courses
";
        return 0;
    }

    sort(results.begin(), results.end(), [](const Result &a, const Result &b) {
        if (a.popularity != b.popularity) {
            return a.popularity > b.popularity;
        }
        return a.course->name < b.course->name;
    });

    cout.setf(ios::fixed);
    cout << setprecision(1);
    for (const auto &r : results) {
        const auto *c = r.course;
        cout << c->name << " by " << c->professor << " (" << c->title << ") - "
             << r.popularity << "% (" << c->enrolled << "/" << c->capacity << ")
";
    }
    return 0;
}
